.changeset/selfish-coats-shout.md

```diff
@@ -0,0 +1,58 @@
+---
+'@backstage/config-loader': minor
+---
+
+Adds a new `deepVisibility` schema keyword that sets child visibility recursively to the defined value, respecting preexisting values or child `deepVisibility`.
+
+Example usage:
+
+```ts
+export interface Config {
+  /**
+   * Enforces a default of `secret` instead of `backend` for this object.
+   * @deepVisibility secret
+   *
+   * `@deepVisibility` only changes the child visibility, you still need to set visibility.
+   * @visibility secret
```

Comment:
I find this a bit strange for two reasons, both because I would assume deep visibility would apply at its own level too, but also because deep visibility wouldn't generally be useful on its own level anyway as you'd apply it to collections. I guess it becomes more important when we consider unions?

Either way, would it make sense to have deep visibility apply at its own level too?

Reply:
Yeah, I've been flipping back and forth on whether this is worth including at the same level. Would we have any reason to want to hide _keys_ of the children of secret objects? Looking at the config reader, that seems to not be the case, https://github.com/backstage/backstage/blob/master/packages/config-loader/src/schema/filtering.ts#L51, but would be the primary use case for having `deepVisibility` only apply to children. I think that would be a large breaking change though.

Unions would definitely be a weirdness, if we switch to applying to self as well, I think that would work there as well.

I'm good to adjust this back, I think it makes more sense to have it apply to both.